integration-tests/tests/src/tests/counter.ts

```diff
@@ -0,0 +1,994 @@
+        });
+      });
+      expectCounter(counter);
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          counter.decrement(1.1);
+        });
+      }).to.throw("Expected 'by' to be an integer, got a floating point number");
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          counter.decrement(NaN);
+        });
+      }).to.throw("Expected 'by' to be an integer, got NaN");
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          // @ts-expect-error Testing incorrect type.
+          counter.decrement(new Number(1));
+        });
+      }).to.throw("Expected 'by' to be an integer, got an instance of Number");
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          // @ts-expect-error Testing incorrect type.
+          counter.decrement("1");
+        });
+      }).to.throw("Expected 'by' to be an integer, got a string");
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          // @ts-expect-error Testing incorrect type.
+          counter.decrement(BigInt(1));
+        });
+      }).to.throw("Expected 'by' to be an integer, got a bigint");
+      expect(counter.value).equals(10);
+    });
+
+    it("throws when setting to non-integer", function (this: RealmContext) {
+      const { counter } = this.realm.write(() => {
+        return this.realm.create<IWithCounter>(WithCounterSchema.name, {
+          counter: 10,
+        });
+      });
+      expectCounter(counter);
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          counter.set(1.1);
+        });
+      }).to.throw("Expected 'value' to be an integer, got a floating point number");
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          counter.set(NaN);
+        });
+      }).to.throw("Expected 'value' to be an integer, got NaN");
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          // @ts-expect-error Testing incorrect type.
+          counter.set(new Number(1));
+        });
+      }).to.throw("Expected 'value' to be an integer, got an instance of Number");
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          // @ts-expect-error Testing incorrect type.
+          counter.set("1");
+        });
+      }).to.throw("Expected 'value' to be an integer, got a string");
+      expect(counter.value).equals(10);
+
+      expect(() => {
+        this.realm.write(() => {
+          // @ts-expect-error Testing incorrect type.
+          counter.set(BigInt(1));
+        });
+      }).to.throw("Expected 'value' to be an integer, got a bigint");
+      expect(counter.value).equals(10);
+    });
```

Comment:
I'm wondering if this could benefit from a loop, to ensure consistency if a case is added in the future:

```
for (const operation of ["increment", "decrement", "set"]) {
```

Reply:
Can take a look at what that refactor would look like 👍

Reply:
Good suggestion Kraen, I've updated it.